### PR TITLE
OSAC-463: Enforce VirtualNetwork child referential integrity via DB triggers

### PR DESCRIPTION
**Pull request by tchughesiv**
## [OSAC-463](https://redhat.atlassian.net/browse/OSAC-463): VirtualNetwork can be deleted while still referenced by Subnet

**Jira:** https://redhat.atlassian.net/browse/OSAC-463
**Story type:** Bug

### Summary

Enforces VirtualNetwork ↔ Subnet/SecurityGroup referential integrity at the database layer using migration 54, following the same trigger pattern as [OSAC-879](https://redhat.atlassian.net/browse/OSAC-879) (migration 52 for Subnet ↔ ComputeInstance).

- **BEFORE UPDATE on `virtual_networks`**: block soft-delete when active Subnets or SecurityGroups reference the VN (`FailedPrecondition` / Z0003)
- **BEFORE INSERT on `subnets` and `security_groups`**: validate the referenced VirtualNetwork exists and is not deleted (`InvalidArgument` / Z0002)

Soft-deleted children no longer block VN deletion — the delete trigger counts only rows with `deletion_timestamp = 'epoch'`.

This PR **supersedes [#586](https://github.com/osac-project/fulfillment-service/pull/586)**. Please close #586 in favor of this PR.

### Changes

- **`internal/database/migrations/54_add_virtual_network_child_ref_triggers.up.sql`**
  - Bidirectional triggers + partial indexes on `data->'spec'->>'virtual_network'`
- **`internal/database/migrations/54_add_virtual_network_child_ref_triggers_test.go`**
  - 13 migration tests covering block/allow paths, soft-delete semantics, and insert validation
- **`internal/servers/private_virtual_networks_server_test.go`**
  - `"Deletion validation"` suite (server-level delete behavior via generic DAO + triggers)
- **`internal/servers/compute_instances_server_test.go`**, **`private_compute_instances_server_test.go`**
  - Seed `test-vnet` before `test-subnet` in BeforeEach (required by insert trigger)

No app-layer `checkNoChildReferences()` guard — `PrivateVirtualNetworksServer.Delete()` delegates to `generic.Delete()` like other resources under [OSAC-879](https://redhat.atlassian.net/browse/OSAC-879).

Public API inherits enforcement via delegation — no changes to `virtual_networks_server.go`.

### Testing

- **Migration tests:** 13 specs in migration 54 test file
- **Server tests:** `"Deletion validation"` suite on private VirtualNetworks server
- **Fixture fix:** compute instance server BeforeEach creates parent VN before subnet

Validation run locally:
- `ginkgo run --focus="Compute instances server|Private compute instances server" internal/servers` — 67/67 pass
- `ginkgo run --focus="Deletion validation" internal/servers` — pass
- `ginkgo run internal/database/migrations --focus="54_add_virtual_network_child_ref_triggers"` — pass

### Acceptance Criteria

- [x] [AC-1](https://redhat.atlassian.net/browse/AC-1): Block delete when Subnets reference the VirtualNetwork (`FailedPrecondition`)
- [x] AC-2: Block delete when SecurityGroups reference the VirtualNetwork (`FailedPrecondition`)
- [x] [AC-3](https://redhat.atlassian.net/browse/AC-3): Allow delete when no active child resources exist; allow after children soft-deleted
- [x] AC-4: Fix at fulfillment-service API level (enforced via DB triggers on generic delete path)
- [x] [AC-5](https://redhat.atlassian.net/browse/AC-5): Prevents the reproducible failure mode from the asciinema recording

```bash
### Test with Subnet
% osac create virtualnetwork -n osac463-vn \
  --network-class 019eb21a-c335-742d-a72b-d099c76be44d \
  --ipv4-cidr 10.99.0.0/16
Created virtual network 'osac463-vn' (ID: 019ebd4e-03fd-7758-a28b-2eec472804a0).

% osac create subnet -n osac463-subnet \
  --virtual-network osac463-vn \
  --ipv4-cidr 10.99.1.0/24
Created subnet 'osac463-subnet' (ID: 019ebd4e-a52c-72ef-bc9f-3fa7543feda7).

% osac delete virtualnetwork osac463-vn
Failed to delete virtualnetwork '019ebd4e-03fd-7758-a28b-2eec472804a0': rpc error: code = FailedPrecondition desc = cannot delete VirtualNetwork '019ebd4e-03fd-7758-a28b-2eec472804a0': 1 Subnet(s) still reference it
Error: 1

### Delete works once references removed
% osac delete subnet osac463-subnet
Deleted subnet '019ebd4e-a52c-72ef-bc9f-3fa7543feda7'.

% osac delete virtualnetwork osac463-vn
Deleted virtualnetwork '019ebd4e-03fd-7758-a28b-2eec472804a0'.
```

### Related

- Precedent: [OSAC-879](https://redhat.atlassian.net/browse/OSAC-879) / migration 52 (Subnet ↔ ComputeInstance triggers)

## Summary by CodeRabbit

* **Tests**
  * Introduced validation tests for virtual network deletion when dependent resources exist
  * Tests confirm deletion fails when subnets or security groups reference the network
  * Tests validate error messages include details about dependent resource types and counts
  * Updated test infrastructure for compute instance operations

### CODE REVIEW

**Comment by tchughesiv** [comment]
Addressed CodeRabbit feedback in b100e1f:

**Pre-merge checks**
- **Title:** Updated to describe the guard behavior (block delete when child refs exist).
- **Docstrings:** Added docs on `Delete()` and expanded `checkNoChildReferences()`.

**Inline comments**
- **Test fixtures:** Removed unnecessary `osac.io/owner-reference` annotations — the guard only checks `spec.virtual_network`.
- **UUID validation:** Declining for now; matches the existing `catalog_item_reference_checker` pattern (no format pre-check, `%q`-quoted CEL filter). Happy to add if we want this consistently across all reference guards.

Focused tests still pass: `ginkgo run --focus="Deletion referential integrity" internal/servers` (6/6).

File: internal/servers/private_virtual_networks_server.go [hunk 1]
```diff
@@ -188,10 +212,57 @@ func (s *PrivateVirtualNetworksServer) Update(ctx context.Context,
...
 	return
 }
 
+// checkNoChildReferences verifies that no Subnets or SecurityGroups reference the given VirtualNetwork.
+func (s *PrivateVirtualNetworksServer) checkNoChildReferences(ctx context.Context, virtualNetworkID string) error {
+	if virtualNetworkID == "" {
+		return nil
+	}
+
+	filter := fmt.Sprintf("this.spec.virtual_network == %q", virtualNetworkID)
```
[the remark this answers is not in the document]
Comment (tchughesiv) [reply internal/servers/private_virtual_networks_server.go]: Declining for now; matches the existing catalog_item_reference_checker pattern (no format pre-check, %q-quoted CEL filter). Happy to add if we want this consistently across all reference guards.

**Comment by tchughesiv** [comment]
@coderabbitai review

**Comment by tchughesiv** [comment]
@coderabbitai full review

File: internal/servers/private_virtual_networks_server.go [hunk 2]
```diff
@@ -186,12 +210,62 @@ func (s *PrivateVirtualNetworksServer) Update(ctx context.Context,
...
+
+	subnetResponse, err := s.subnetDao.List().
+		SetFilter(filter).
+		SetLimit(1).
```
Comment (jhernand) [inline internal/servers/private_virtual_networks_server.go]: If you aren't interested in the items but only on the count you can put 0 here.

File: internal/servers/private_virtual_networks_server.go [hunk 3]
```diff
@@ -186,12 +210,62 @@ func (s *PrivateVirtualNetworksServer) Update(ctx context.Context,
...
+		s.logger.ErrorContext(ctx, "Failed to check subnet references",
+			slog.String("virtual_network_id", virtualNetworkID),
+			slog.Any("error", err))
+		return grpcstatus.Errorf(grpccodes.Internal, "failed to check child references")
+	}
+	if subnetResponse.GetTotal() > 0 {
+		return grpcstatus.Errorf(grpccodes.FailedPrecondition,
+			"cannot delete VirtualNetwork '%s': %d Subnet(s) still reference it",
+			virtualNetworkID, subnetResponse.GetTotal())
+	}
```
Comment (jhernand) [inline internal/servers/private_virtual_networks_server.go]: What happens if a new subnet referencing the virtual network is created after this check but before the delete is sent to the database?

File: internal/servers/private_virtual_networks_server.go [hunk 4]
```diff
@@ -186,12 +210,62 @@ func (s *PrivateVirtualNetworksServer) Update(ctx context.Context,
...
+		s.logger.ErrorContext(ctx, "Failed to check security group references",
+			slog.String("virtual_network_id", virtualNetworkID),
+			slog.Any("error", err))
+		return grpcstatus.Errorf(grpccodes.Internal, "failed to check child references")
+	}
+	if sgResponse.GetTotal() > 0 {
+		return grpcstatus.Errorf(grpccodes.FailedPrecondition,
+			"cannot delete VirtualNetwork '%s': %d SecurityGroup(s) still reference it",
+			virtualNetworkID, sgResponse.GetTotal())
+	}
```
Comment (jhernand) [inline internal/servers/private_virtual_networks_server.go]: Same, what happens if a security group referencing this virtual network is created after this check but before the delete?

File: internal/servers/private_virtual_networks_server.go (hunk 2 above)
Comment (tchughesiv) [reply internal/servers/private_virtual_networks_server.go]: Thanks... we only need `GetTotal()`, not the returned items. I kept [SetLimit(1)](https://github.com/osac-project/fulfillment-service/blob/b100e1f9159a9e7740a53d856045b3f1dcdc24d6/internal/servers/private_virtual_networks_server.go#L234-L237) because in our DAO [SetLimit(0) maps to defaultLimit (100)](https://github.com/osac-project/fulfillment-service/blob/main/internal/database/dao/generic_dao_list.go#L146-L156), not “return zero rows” — see also the test [“Interprets zero limit as requesting the default number of items”](https://github.com/osac-project/fulfillment-service/blob/main/internal/database/dao/generic_dao_test.go#L1168-L1174).

[SetLimit(1)](https://github.com/osac-project/fulfillment-service/blob/b100e1f9159a9e7740a53d856045b3f1dcdc24d6/internal/servers/private_virtual_networks_server.go#L234-L237) is the minimal row fetch today (the [count(*)](https://github.com/osac-project/fulfillment-service/blob/main/internal/database/dao/generic_dao_list.go#L89-L108) query always runs regardless). Same pattern as [daoReferenceChecker.hasReference()](https://github.com/osac-project/fulfillment-service/blob/main/internal/servers/catalog_item_reference_checker.go#L36-L45).

If we want count-only behavior repo-wide, I can follow up with a small DAO change (skip the SELECT when only the count is needed) — happy to do that in a separate commit on this branch or a follow-up PR, whichever the team prefers.

File: internal/servers/private_virtual_networks_server.go (hunk 3 above)
Comment (tchughesiv) [reply internal/servers/private_virtual_networks_server.go]: Same pattern as [PrivateClusterCatalogItemsServer.Delete](https://github.com/osac-project/fulfillment-service/blob/main/internal/servers/private_cluster_catalog_items_server.go#L152-L167) + [daoReferenceChecker.hasReference()](https://github.com/osac-project/fulfillment-service/blob/main/internal/servers/catalog_item_reference_checker.go#L36-L45): check then delete inside one RPC handler, both DAO calls using the transaction from [TxInterceptor.UnaryServer](https://github.com/osac-project/fulfillment-service/blob/main/internal/database/database_tx_interceptor.go#L76-L125) ([List reads tx from context](https://github.com/osac-project/fulfillment-service/blob/main/internal/database/dao/generic_dao_list.go#L57-L63)).

Under PostgreSQL READ COMMITTED, a concurrent create that commits between the [List count](https://github.com/osac-project/fulfillment-service/blob/b100e1f9159a9e7740a53d856045b3f1dcdc24d6/internal/servers/private_virtual_networks_server.go#L234-L248) and [generic.Delete](https://github.com/osac-project/fulfillment-service/blob/b100e1f9159a9e7740a53d856045b3f1dcdc24d6/internal/servers/private_virtual_networks_server.go#L213-L222) could theoretically slip through... same tradeoff we already accept for catalog-item delete.

OSAC-463 is the application-layer guard for the common case (delete VN while children still exist). Hard guarantees would need DB triggers like [migration 52 (subnet ↔ compute instance)](https://github.com/osac-project/fulfillment-service/blob/main/internal/database/migrations/52_add_subnet_compute_instance_ref_triggers.up.sql#L14-L24). We should probably treat that as a separate design decision, not part of this PR unless you want it scoped in.

File: internal/servers/private_virtual_networks_server.go (hunk 4 above)
Comment (tchughesiv) [reply internal/servers/private_virtual_networks_server.go]: same

File: internal/servers/private_virtual_networks_server.go (hunk 3 above)
Comment (tchughesiv) [reply internal/servers/private_virtual_networks_server.go]: looking closer at https://github.com/osac-project/fulfillment-service/pull/646 ... it used DB triggers. i'll add a new commit to this PR implementing similar approach and we can decide if we want to keep it or not.

Comment (tchughesiv) [reply internal/servers/private_virtual_networks_server.go]: adopted the [OSAC-879 / migration 52](https://github.com/osac-project/fulfillment-service/blob/main/internal/database/migrations/52_add_subnet_compute_instance_ref_triggers.up.sql#L14-L24) pattern in [migration 54](https://github.com/osac-project/fulfillment-service/blob/97546829fa3d6b6df54828928391e2e4ab22c88f/internal/database/migrations/54_add_virtual_network_child_ref_triggers.up.sql): bidirectional triggers with FOR SHARE on child insert. removed the app-layer checkNoChildReferences... enforcement is DB-only, same as subnet ↔ compute instance.

File: internal/servers/private_virtual_networks_server.go (hunk 2 above)
Comment (tchughesiv) [reply internal/servers/private_virtual_networks_server.go]: no longer relevant if we keep DB trigger approach

File: internal/database/migrations/55_add_virtual_network_child_ref_triggers_test.go [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+				select
+					count(*)
+				from
+					pg_indexes
+				where
+					indexname = $1
+			`, indexName).Scan(&count)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(count).To(Equal(1))
+		}
```
[the remark this answers is not in the document]
Comment (tchughesiv) [reply internal/database/migrations/55_add_virtual_network_child_ref_triggers_test.go:95]: Done in c5a48e0 — the index test now queries `pg_indexes.indexdef` and asserts both the JSONB expression and the partial-index predicate (using PostgreSQL's literal `'1970-01-01 00:00:00+00'` form for `epoch`, which is what `indexdef` reports).

File: internal/database/migrations/55_add_virtual_network_child_ref_triggers_test.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+		_, err = conn.Exec(ctx,
+			`insert into security_groups (id, tenant, data)
+			 values ('sg-ref', 'test-tenant', $1::jsonb)`,
+			`{"spec":{"virtual_network":"vn-sg-deleted"}}`)
+		Expect(err).To(HaveOccurred())
+		var pgErr *pgconn.PgError
+		Expect(errors.As(err, &pgErr)).To(BeTrue())
+		Expect(pgErr.Code).To(Equal("Z0002"))
+		Expect(pgErr.Message).To(ContainSubstring("vn-sg-deleted"))
+	})
```
[the remark this answers is not in the document]
Comment (tchughesiv) [reply internal/database/migrations/55_add_virtual_network_child_ref_triggers_test.go:336]: Done in c5a48e0 — added `Prevents creating a security group referencing a non-existent virtual network`, mirroring the existing subnet test for the `no-such-vn` / Z0002 path.

File: internal/servers/private_virtual_networks_server.go (hunk 2 above)
Comment (jhernand) [reply internal/servers/private_virtual_networks_server.go]: You are correct; it is a negative value that means "count only": https://github.com/osac-project/fulfillment-service/blob/dc987c9189e68567c1eccfb1c946bdb9de76fef5/internal/database/dao/generic_dao_test.go#L1159-L1166 . So use -1, for example.

I opened https://redhat.atlassian.net/browse/OSAC-1541 to improve that.

**Review by jhernand — APPROVED** [review]
